service/src/amqp.ts

```diff
@@ -69,7 +80,331 @@ export class Amqp extends EventEmitter implements Client.Transport {
+                }
+              },
+              (callback) => {
+                /*Codes_SRS_NODE_IOTHUB_SERVICE_AMQP_16_022: [The `disconnect` method shall detach the C2D feedback receiver link if it is attached.]*/
+                if (this._feedbackReceiver) {
+                  let tmpFeedbackReceiver = this._feedbackReceiver;
+                  this._feedbackReceiver = undefined;
+
+                  if (err) {
+                    tmpFeedbackReceiver.forceDetach(err);
+                    tmpFeedbackReceiver.removeListener('error', this._feedbackErrorListener);
+                    callback();
+                  } else {
+                    tmpFeedbackReceiver.detach((detachErr) => {
+                      if (detachErr) {
+                        debug('error detaching the message feedback link: ' + detachErr.toString());
+                      } else {
+                        debug('D2C link detached');
+                      }
+                      tmpFeedbackReceiver.removeListener('error', this._feedbackErrorListener);
+                      if (!finalError && detachErr) {
+                        finalError = translateError('error while detaching the message feedback link when disconnecting', detachErr);
+                      }
+                      callback();
+                    });
+                  }
+                } else {
+                  callback();
+                }
+              },
+              /*Codes_SRS_NODE_IOTHUB_SERVICE_AMQP_16_023: [The `disconnect` method shall detach the file notification receiver link if it is attached.]*/
+              (callback) => {
+                if (this._fileNotificationReceiver) {
+                  let tmpFileNotificationReceiver = this._fileNotificationReceiver;
+                  this._fileNotificationReceiver = undefined;
+
+                  if (err) {
+                    tmpFileNotificationReceiver.forceDetach(err);
+                    tmpFileNotificationReceiver.removeListener('error', this._fileNotificationErrorListener);
+                    callback();
+                  } else {
+                    tmpFileNotificationReceiver.detach((detachErr) => {
+                      if (detachErr) {
+                        debug('error detaching the file upload notification link: ' + detachErr.toString());
+                      } else {
+                        debug('D2C link detached');
+                      }
+                      tmpFileNotificationReceiver.removeListener('error', this._fileNotificationErrorListener);
+                      if (!finalError && detachErr) {
+                        finalError = translateError('error while detaching the file upload notification link when disconnecting', detachErr);
+                      }
+                      callback();
+                    });
+                  }
+                } else {
+                  callback();
+                }
+              },
+              (callback) => {
```

Comment:
>(callback) => { [](start = 14, length = 15)

Do you want to clear the put token renewal timer?

Reply:
this is done on the _onExit of the authenticated state

---
In reply to: [147610067](https://github.com/Azure/azure-iot-sdk-node/pull/127#discussion_r147610067) [](ancestors = 147610067)

Reply:
I have moved the call to setTimeout in the _onEnter too to make this more symmetrical

---
In reply to: [148421117](https://github.com/Azure/azure-iot-sdk-node/pull/127#discussion_r148421117) [](ancestors = 148421117,147610067)